js/piwik.js

```diff
@@ -2894,6 +2894,14 @@ if (typeof window.Matomo !== 'object') {
                 }
             }
 
+            function heartBeatOnVisible() {
+                //come back to original tab
+                if (document.visibilityState === 'visible' && hadWindowMinimalFocusToConsiderViewed()) {
```

Comment:
I haven't tested it, but usually we send the `ping` event on "blur" meaning when it becomes hidden. I wonder if we need to test here for 

```js
if (documentAlias.visibilityState === 'hidden' && hadWindowMinimalFocusToConsiderViewed()) {
    heartBeatPingIfActivityAlias();
} else if (documentAlias.visibilityState === 'visible') {
    timeWindowLastFocused = new Date().getTime();
}
```

seems something like this would maybe replicate the focus/blur event behaviour? Again, I haven't tested though. Just wanted to give you some quick feedback. 

Also we usually don't use `document` directly but `documentAlias`. This way the JS minifier can optimise these calls using a shorter variable.

Reply:
@tsteur you are right, they replicate the blur behavior.